File: mRemoteV1/Connection/Connection.Protocol.IntegratedProgram.vb
Imports System.Threading
Imports mRemote3G.App
Imports mRemote3G.Messages
Imports mRemote3G.My
Imports mRemote3G.Tools

Namespace Connection

    Namespace Protocol
        Public Class IntegratedProgram
            Inherits Base

#Region "Public Methods"

            Public Overrides Function SetProps() As Boolean
                If InterfaceControl.Info IsNot Nothing Then
                    _externalTool = Runtime.GetExtAppByName(InterfaceControl.Info.ExtApp)
                    _externalTool.ConnectionInfo = InterfaceControl.Info
                End If

                Return MyBase.SetProps()
            End Function

            Public Overrides Function Connect() As Boolean
                Try
                    If _externalTool.TryIntegrate = False Then
                        _externalTool.Start(InterfaceControl.Info)
                        Close()
                        Return Nothing
                    End If

                    _process = New Process()

                    With _process.StartInfo
                        .UseShellExecute = True
                        .FileName = _externalTool.ParseArguments(_externalTool.FileName)
                        .Arguments = _externalTool.ParseArguments(_externalTool.Arguments)
                    End With

                    _process.EnableRaisingEvents = True
                    AddHandler _process.Exited, AddressOf ProcessExited

                    _process.Start()
                    _process.WaitForInputIdle(Settings.MaxPuttyWaitTime*1000)

                    Dim startTicks As Integer = Environment.TickCount
                    While _handle.ToInt32 = 0 And Environment.TickCount < startTicks + (Settings.MaxPuttyWaitTime*1000)
                        _process.Refresh()
                        If Not _process.MainWindowTitle = "Default IME" Then _handle = _process.MainWindowHandle
                        If _handle.ToInt32 = 0 Then Thread.Sleep(0)
                    End While

                    NativeMethods.SetParent(_handle, InterfaceControl.Handle)

                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, Language.Language.strIntAppStuff, True)

                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, String.Format(Language.Language.strIntAppHandle, _handle.ToString), True)
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, String.Format(Language.Language.strIntAppTitle, _process.MainWindowTitle), True)
                   App.Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, String.Format(Language.Language.strIntAppParentHandle, InterfaceControl.Parent.Handle.ToString), True)

                    Resize(Me, New EventArgs)

                    MyBase.Connect()
                    Return True
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strIntAppConnectionFailed, ex)
                    Return False
                End Try
            End Function

            Public Overrides Sub Focus()
                Try
                    If ConnectionWindow.InTabDrag Then Return
                    NativeMethods.SetForegroundWindow(_handle)
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strIntAppFocusFailed, ex, , True)
                End Try
            End Sub

            Public Overrides Sub Resize(sender As Object, e As EventArgs)
                Try
                    If InterfaceControl.Size = Size.Empty Then Return
                    NativeMethods.MoveWindow(_handle, -SystemInformation.FrameBorderSize.Width,
                                             (SystemInformation.CaptionHeight + SystemInformation.FrameBorderSize.Height),
                                             InterfaceControl.Width + (SystemInformation.FrameBorderSize.Width * 2),
                                             InterfaceControl.Height + SystemInformation.CaptionHeight +
                                             (SystemInformation.FrameBorderSize.Height * 2), True)
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddExceptionMessage(Language.Language.strIntAppResizeFailed, ex, , True)
                End Try
            End Sub

            Public Overrides Sub Close()
                Try
                    If Not _process.HasExited Then _process.Kill()
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddExceptionMessage(Language.Language.strIntAppKillFailed, ex, , True)
                End Try

                Try
                    If Not _process.HasExited Then _process.Dispose()
                Catch ex As Exception
                   App.Runtime.MessageCollector.AddExceptionMessage(Language.Language.strIntAppDisposeFailed, ex, , True)
                End Try

                MyBase.Close()
            End Sub

#End Region

#Region "Private Fields"

            Private _externalTool As ExternalTool
            Private _handle As IntPtr
            Private _process As Process

#End Region

#Region "Private Methods"

            Private Sub ProcessExited(sender As Object, e As EventArgs)
                Event_Closed(Me)
            End Sub

#End Region

#Region "Enumerations"

            Public Enum Defaults
                Port = 0
            End Enum

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1/CS/Connection/Connection.Protocol.IntegratedProgram.cs
using System;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Threading;
using mRemoteNG.App;
using mRemoteNG.Tools;


namespace mRemoteNG.Connection.Protocol
{
	public class IntegratedProgram : Base
	{
        #region Public Methods
		public override bool SetProps()
		{
			if (InterfaceControl.Info != null)
			{
				_externalTool = Runtime.GetExtAppByName(System.Convert.ToString(InterfaceControl.Info.ExtApp));
				_externalTool.ConnectionInfo = InterfaceControl.Info;
			}
					
			return base.SetProps();
		}
				
		public override bool Connect()
		{
			try
			{
				if (_externalTool.TryIntegrate == false)
				{
					_externalTool.Start(InterfaceControl.Info);
					Close();
					return false;
				}
						
				_process = new Process();
						
				_process.StartInfo.UseShellExecute = true;
				_process.StartInfo.FileName = _externalTool.ParseArguments(_externalTool.FileName);
				_process.StartInfo.Arguments = _externalTool.ParseArguments(_externalTool.Arguments);
						
				_process.EnableRaisingEvents = true;
				_process.Exited += ProcessExited;
						
				_process.Start();
				_process.WaitForInputIdle(System.Convert.ToInt32(My.Settings.Default.MaxPuttyWaitTime * 1000));
						
				int startTicks = Environment.TickCount;
				while (_handle.ToInt32() == 0 & Environment.TickCount < startTicks + (My.Settings.Default.MaxPuttyWaitTime * 1000))
				{
					_process.Refresh();
					if (!(_process.MainWindowTitle == "Default IME"))
					{
						_handle = _process.MainWindowHandle;
					}
					if (_handle.ToInt32() == 0)
					{
						Thread.Sleep(0);
					}
				}
						
				Native.SetParent(_handle, InterfaceControl.Handle);
				Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, My.Language.strIntAppStuff, true);
				Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, string.Format(My.Language.strIntAppHandle, _handle.ToString()), true);
				Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, string.Format(My.Language.strIntAppTitle, _process.MainWindowTitle), true);
				Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, string.Format(My.Language.strIntAppParentHandle, InterfaceControl.Parent.Handle.ToString()), true);
						
				Resize(this, new EventArgs());
				base.Connect();
				return true;
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(My.Language.strIntAppConnectionFailed, ex);
				return false;
			}
		}
				
		public override void Focus()
		{
			try
			{
				if (ConnectionWindow.InTabDrag)
				{
					return ;
				}
				Native.SetForegroundWindow(_handle);
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(message: My.Language.strIntAppFocusFailed, ex: ex, logOnly: true);
			}
		}
				
		public override void Resize(object sender, EventArgs e)
		{
			try
			{
				if (InterfaceControl.Size == Size.Empty)
				{
					return ;
				}
                Native.MoveWindow(_handle, System.Convert.ToInt32(-SystemInformation.FrameBorderSize.Width), System.Convert.ToInt32(-(SystemInformation.CaptionHeight + SystemInformation.FrameBorderSize.Height)), InterfaceControl.Width + (SystemInformation.FrameBorderSize.Width * 2), InterfaceControl.Height + SystemInformation.CaptionHeight + (SystemInformation.FrameBorderSize.Height * 2), true);
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(message: My.Language.strIntAppResizeFailed, ex: ex, logOnly: true);
			}
		}
				
		public override void Close()
		{
			try
			{
				if (!_process.HasExited)
				{
					_process.Kill();
				}
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(message: My.Language.strIntAppKillFailed, ex: ex, logOnly: true);
			}
					
			try
			{
				if (!_process.HasExited)
				{
					_process.Dispose();
				}
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(message: My.Language.strIntAppDisposeFailed, ex: ex, logOnly: true);
			}
					
			base.Close();
		}
        #endregion
				
        #region Private Fields
		private ExternalTool _externalTool;
		private IntPtr _handle;
		private Process _process;
        #endregion
				
        #region Private Methods
		private void ProcessExited(object sender, EventArgs e)
		{
			Event_Closed(this);
		}
        #endregion
				
        #region Enumerations
		public enum Defaults
		{
			Port = 0
		}
        #endregion
	}
}
